File: bindings/php/examples/basic_usage.php
<?php

declare(strict_types=1);

/**
 * Ainos PHP SDK - Basic Usage Example
 *
 * This example demonstrates the basic usage of the Ainos client:
 * - Connecting to the server
 * - Checking health and status
 * - Listing available models
 * - Performing synchronous inference
 * - Managing context
 * - Disconnecting
 *
 * Usage:
 *   export AINOS_TOKEN=your-token-here
 *   php examples/basic_usage.php
 *
 * Or set the token directly in the script.
 */

// Load the SDK via Composer autoloader
require_once __DIR__ . '/../vendor/autoload.php';

use Ainos\AinosClient;
use Ainos\Authentication;
use Ainos\Parameters;
use Ainos\Utils;

// ----------------------------------------------------------------
// 1. Configuration
// ----------------------------------------------------------------

// Option A: Configure from environment variables
// $client = AinosClient::fromEnvironment();

// Option B: Configure directly
$token = \getenv('AINOS_TOKEN') ?: 'your-token-here';
$host = '127.0.0.1';
$port = 9500;

$auth = new Authentication($token);

$client = new AinosClient(
    $auth,
    $host,
    $port,
    [
        'timeout' => 30.0,
        'max_retries' => 3,
        'retry_delay' => 1.0,
    ]
);

// ----------------------------------------------------------------
// 2. Connection
// ----------------------------------------------------------------

echo "Connecting to Ainos server at {$host}:{$port}...\n";

try {
    $client->connect();
    echo "Connected successfully.\n\n";
} catch (\Ainos\ConnectionException $e) {
    echo "Connection failed: {$e->getMessage()}\n";
    exit(1);
}

// ----------------------------------------------------------------
// 3. Server Health
// ----------------------------------------------------------------

echo "--- Server Health ---\n";

try {
    $health = $client->health();

    echo "Status: {$health->status}\n";
    echo "Version: {$health->version}\n";
    echo "Uptime: {$health->getUptimeFormatted()}\n";
    echo "Active Connections: {$health->activeConnections}\n";
    echo "Healthy: " . ($health->isHealthy() ? 'Yes' : 'No') . "\n\n";
} catch (\Ainos\AinosException $e) {
    echo "Health check failed: {$e->getMessage()}\n\n";
}

// ----------------------------------------------------------------
// 4. Server Status
// ----------------------------------------------------------------

echo "--- Server Status ---\n";

try {
    $status = $client->status();

    echo "Version: {$status->version}\n";
    echo "Uptime: {$status->uptime} seconds\n";
    echo "Total Requests: {$status->totalRequests}\n";
    echo "Active Models: " . \implode(', ', $status->activeModels) . "\n";
    echo "Active Connections: {$status->activeConnections}\n\n";
} catch (\Ainos\AinosException $e) {
    echo "Status check failed: {$e->getMessage()}\n\n";
}

// ----------------------------------------------------------------
// 5. Model List
// ----------------------------------------------------------------

echo "--- Available Models ---\n";

try {
    $models = $client->modelList();

    echo "Total Models: {$models->total}\n";
    echo "Loaded Models: {$models->loadedCount}\n";
    echo "Total Size: " . Utils::formatBytes($models->totalSize) . "\n\n";

    foreach ($models->models as $model) {
        echo "  - {$model->name}\n";
        echo "    ID: {$model->id}\n";
        echo "    Size: {$model->getSizeFormatted()}\n";
        echo "    Loaded: " . ($model->loaded ? 'Yes' : 'No') . "\n";
        echo "    Version: " . ($model->version ?? 'N/A') . "\n\n";
    }
} catch (\Ainos\AinosException $e) {
    echo "Model list failed: {$e->getMessage()}\n\n";
}

// ----------------------------------------------------------------
// 6. Synchronous Inference
// ----------------------------------------------------------------

echo "--- Synchronous Inference ---\n";

$modelName = 'gpt-3.5-turbo';
$prompt = 'Explain what a TCP socket is in one sentence.';

echo "Model: {$modelName}\n";
echo "Prompt: {$prompt}\n\n";

try {
    $timer = Utils::microtimeFloat();

    $response = $client->infer(
        $modelName,
        $prompt,
        new Parameters(
            temperature: 0.7,
            maxTokens: 100,
            topP: 0.9,
            stop: ['.', '!', '?'],
        ),
    );

    $duration = Utils::microtimeFloat() - $timer;

    echo "Response ID: {$response->id}\n";
    echo "Generated Text:\n{$response->getText()}\n\n";
    echo "Usage:\n";
    echo "  Prompt Tokens: {$response->usage->promptTokens}\n";
    echo "  Completion Tokens: {$response->usage->completionTokens}\n";
    echo "  Total Tokens: {$response->usage->totalTokens}\n";
    echo "  Duration: " . \round($duration, 3) . "s\n\n";
} catch (\Ainos\AinosException $e) {
    echo "Inference failed: {$e->getMessage()}\n\n";
}

// ----------------------------------------------------------------
// 7. Context Management
// ----------------------------------------------------------------

echo "--- Context Management ---\n";

try {
    // Store a context value
    $entry = $client->contextStore(
        'user-session-123',
        ['name' => 'Alice', 'preferences' => ['theme' => 'dark']],
        3600, // 1 hour TTL
    );

    echo "Context Stored:\n";
    echo "  Key: {$entry->key}\n";
    echo "  TTL: {$entry->ttl}s\n";
    echo "  Expires: " . \date('Y-m-d H:i:s', $entry->expiresAt ?? 0) . "\n";

    // Retrieve the context value
    $retrieved = $client->contextRetrieve('user-session-123');

    if ($retrieved !== null) {
        echo "Context Retrieved:\n";
        echo "  Key: {$retrieved->key}\n";
        echo "  Value: " . \json_encode($retrieved->value) . "\n";
        echo "  Expired: " . ($retrieved->isExpired() ? 'Yes' : 'No') . "\n\n";
    } else {
        echo "Context not found.\n\n";
    }
} catch (\Ainos\AinosException $e) {
    echo "Context operation failed: {$e->getMessage()}\n\n";
}

// ----------------------------------------------------------------
// 8. Client Statistics
// ----------------------------------------------------------------

echo "--- Client Statistics ---\n";

$stats = $client->getStats();

echo "Total Requests: {$stats['total_requests']}\n";
echo "Successful Requests: {$stats['successful_requests']}\n";
echo "Failed Requests: {$stats['failed_requests']}\n";
echo "Token Preview: {$stats['auth_token_preview']}\n";

$transportStats = $stats['transport'];
echo "Bytes Sent: {$transportStats['bytes_sent']}\n";
echo "Bytes Received: {$transportStats['bytes_received']}\n\n";

// ----------------------------------------------------------------
// 9. Cleanup
// ----------------------------------------------------------------

echo "Disconnecting...\n";
$client->disconnect();
echo "Done.\n";

File: bindings/python/examples/basic_usage.py
"""
Ainos SDK - Basic Usage Example
================================

Demonstrates the core functionality of the Ainos SDK:
- Creating a client
- Checking daemon health
- Running non-streaming inference
- Listing models
- Using the context store
- Error handling

Prerequisites:
    - The Ainos daemon must be running on localhost:9500
    - An auth token may be required (set AINOS_AUTH_TOKEN env var)

Usage:
    python examples/basic_usage.py
"""

from __future__ import annotations

import asyncio
import logging
import os
import sys
from typing import Any

# Add the parent directory to the path so we can import ainos
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from ainos import AinosClient, setup_logging
from ainos.errors import (
    AinosError,
    ConnectionError,
    ModelNotLoadedError,
)
from ainos.types import (
    HealthStatus,
    InferenceResponse,
    ModelInfo,
    SystemStatus,
)


async def main() -> None:
    """Run the basic usage example."""
    # Set up logging so we can see what's happening
    setup_logging(level=logging.INFO)

    # Get auth token from environment (or use a default for local dev)
    auth_token: str | None = os.environ.get("AINOS_AUTH_TOKEN")

    print("=" * 60)
    print("Ainos SDK - Basic Usage Example")
    print("=" * 60)

    # ------------------------------------------------------------------
    # 1. Create and connect the client
    # ------------------------------------------------------------------
    print("\n[1] Connecting to Ainos daemon...")

    try:
        client: AinosClient = AinosClient(
            host="127.0.0.1",
            port=9500,
            auth_token=auth_token,
            connect_timeout=5.0,
            request_timeout=30.0,
        )
        await client.connect()
        print(f"    Connected: {client.connected}")
    except ConnectionError as exc:
        print(f"    FAILED: {exc}")
        print("    Is the Ainos daemon running on localhost:9500?")
        sys.exit(1)

    # ------------------------------------------------------------------
    # 2. Health check
    # ------------------------------------------------------------------
    print("\n[2] Checking daemon health...")

    health: HealthStatus = await client.health()
    print(f"    Healthy: {health.healthy}")
    print(f"    Version: {health.version or 'unknown'}")
    print(f"    Status:  {health.status}")
    print(f"    Uptime:  {health.uptime_seconds:.1f}s")
    print(f"    Active models: {health.active_models}")

    # ------------------------------------------------------------------
    # 3. System status
    # ------------------------------------------------------------------
    print("\n[3] Fetching system status...")

    try:
        status: SystemStatus = await client.status()
        print(f"    Daemon version: {status.version}")
        print(f"    Uptime: {status.uptime_seconds:.1f}s")
        print(f"    Memory: {status.memory_used_mb:.0f}MB / {status.memory_total_mb:.0f}MB")
        print(f"    CPU: {status.cpu_usage_percent:.1f}%")
        if status.gpu_usage_percent is not None:
            print(f"    GPU: {status.gpu_usage_percent:.1f}%")
        print(f"    Active requests: {status.active_requests}")
        print(f"    Queued requests: {status.queued_requests}")
        if status.errors:
            print(f"    Errors: {status.errors}")
    except Exception as exc:
        print(f"    Failed to fetch status: {exc}")

    # ------------------------------------------------------------------
    # 4. List models
    # ------------------------------------------------------------------
    print("\n[4] Listing models...")

    try:
        models: list[ModelInfo] = await client.model_list()
        if models:
            print(f"    Found {len(models)} model(s):")
            for i, model in enumerate(models):
                print(f"      [{i + 1}] {model.name} (id={model.id})")
                print(f"           Status: {model.status}")
                print(f"           Backend: {model.backend}")
                print(f"           Device: {model.device}")
                print(f"           Context: {model.context_length} tokens")
        else:
            print("    No models found. Load one with model_load().")
    except Exception as exc:
        print(f"    Failed to list models: {exc}")

    # ------------------------------------------------------------------
    # 5. Non-streaming inference
    # ------------------------------------------------------------------
    print("\n[5] Running non-streaming inference...")

    # Find the first loaded model
    loaded_models: list[ModelInfo] = [m for m in models if m.status == "loaded"] if models else []

    if loaded_models:
        model_name: str = loaded_models[0].id
        print(f"    Using model: {model_name}")

        try:
            response: InferenceResponse = await client.infer(
                model=model_name,
                prompt="What is the meaning of life? Answer in one sentence.",
                max_tokens=50,
                temperature=0.8,
            )
            print(f"    Response: {response.text}")
            print(f"    Finish reason: {response.finish_reason}")
            if response.usage:
                print(f"    Tokens: {response.usage.prompt_tokens} prompt + "
                      f"{response.usage.completion_tokens} completion = "
                      f"{response.usage.total_tokens} total")
        except ModelNotLoadedError as exc:
            print(f"    Model not loaded: {exc}")
        except AinosError as exc:
            print(f"    Inference failed: {exc}")
    else:
        print("    No loaded models available for inference.")
        print("    Use model_management.py example to load a model first.")

    # ------------------------------------------------------------------
    # 6. Context store
    # ------------------------------------------------------------------
    print("\n[6] Using the context store...")

    try:
        # Store a value
        stored: bool = await client.context_store("example_key", {"hello": "world"}, ttl=60)
        print(f"    Stored value: {stored}")

        # Retrieve the value
        value: Any = await client.context_retrieve("example_key")
        print(f"    Retrieved value: {value}")

        # Retrieve a non-existent key
        missing: Any = await client.context_retrieve("non_existent_key")
        print(f"    Non-existent key: {missing}")
    except Exception as exc:
        print(f"    Context store operation failed: {exc}")

    # ------------------------------------------------------------------
    # 7. Disconnect
    # ------------------------------------------------------------------
    print("\n[7] Disconnecting...")

    await client.disconnect()
    print(f"    Connected: {client.connected}")
    print()

    print("=" * 60)
    print("Example completed successfully!")
    print("=" * 60)


if __name__ == "__main__":
    asyncio.run(main())